render/host.go

```diff
@@ -62,18 +59,73 @@ func MapX2Host(n report.Node, _ report.Networks) report.Nodes {
 	return result
 }
 
-// MapEndpoint2Host takes nodes from the endpoint topology and produces
+// endpoints2Hosts takes nodes from the endpoint topology and produces
 // host nodes or pseudo nodes.
-func MapEndpoint2Host(n report.Node, local report.Networks) report.Nodes {
-	// Nodes without a hostid are treated as pseudo nodes
-	hostNodeID, timestamp, ok := n.Latest.LookupEntry(report.HostNodeID)
-	if !ok {
-		return MapEndpoint2Pseudo(n, local)
+type endpoints2Hosts struct {
+}
+
+func (e endpoints2Hosts) Render(rpt report.Report, dct Decorator) report.Nodes {
+	ns := SelectEndpoint.Render(rpt, dct)
+	local := LocalNetworks(rpt)
+
+	var ret = make(report.Nodes)
+	var mapped = map[string]string{} // input node ID -> output node ID
+	for _, n := range ns {
+		// Nodes without a hostid are treated as pseudo nodes
+		hostNodeID, timestamp, ok := n.Latest.LookupEntry(report.HostNodeID)
+		if !ok {
+			id, ok := pseudoNodeID(n, local)
+			if !ok {
+				continue
+			}
+			addToResults(n, id, ret, mapped, func() report.Node {
+				return report.MakeNode(id).WithTopology(Pseudo)
+			})
+		} else {
+			id := report.MakeHostNodeID(report.ExtractHostID(n))
+			addToResults(n, id, ret, mapped, func() report.Node {
+				return report.MakeNode(id).WithTopology(report.Host).
+					WithLatest(report.HostNodeID, timestamp, hostNodeID)
+			})
+		}
 	}
+	fixupAdjancencies(ns, ret, mapped)
+	return ret
+}
+
+// Add Node M to the result set ret under id, creating a new result
+// node if not already there, and updating the old-id to new-id mapping
+// Note we do not update any counters for child topologies here
+func addToResults(m report.Node, id string, ret report.Nodes, mapped map[string]string, create func() report.Node) {
+	result, exists := ret[id]
+	if !exists {
+		result = create()
+	}
+	result.Children = result.Children.Add(m)
+	result.Children = result.Children.Merge(m.Children)
+	ret[result.ID] = result
+	mapped[m.ID] = result.ID
+}
```

Comment:
There are a few too many ids here for my liking. AFAICT we expect `result.ID == id` here, right? So perhaps
1. always use `id`
2. pass `id` to `create()`
3. assert that the `result.ID` returned by `create()` is `id`

The second rather clutters call sites though, since they all use closures which already have access to the id.

Reply:
Actually, doing the 2nd would allow us to introduce a `NewPseudoNode(id string)` function, thus eliminating the closure in three of the `addToResults` call sites.